File: crates/gcode/src/bm25_health.rs
use postgres::Client;
use serde::Serialize;

use crate::postgres_errors;

pub const REPAIR_COMMAND: &str = "gobby postgres repair-code-index";
pub const REQUIRED_INDEXES: [&str; 2] = [
    "public.code_symbols_search_bm25",
    "public.code_content_search_bm25",
];

const CORRUPTION_SQLSTATES: [&str; 3] = ["XX000", "XX001", "XX002"];

#[derive(Debug, Serialize)]
pub struct Bm25Health {
    pub healthy: bool,
    pub repair_command: &'static str,
    pub indexes: Vec<Bm25IndexHealth>,
}

#[derive(Debug, Serialize)]
pub struct Bm25IndexHealth {
    pub name: &'static str,
    pub state: &'static str,
    pub repaired: bool,
    pub checks: Vec<Bm25IndexCheck>,
    pub error: Option<String>,
}

#[derive(Debug, Serialize)]
pub struct Bm25IndexCheck {
    pub name: String,
    pub passed: bool,
    pub details: Option<String>,
}

pub fn verify(conn: &mut Client) -> Bm25Health {
    let indexes = REQUIRED_INDEXES
        .into_iter()
        .map(|name| verify_index(conn, name))
        .collect::<Vec<_>>();
    Bm25Health {
        healthy: indexes.iter().all(|index| index.state == "healthy"),
        repair_command: REPAIR_COMMAND,
        indexes,
    }
}

impl Bm25Health {
    pub fn render_text(&self) -> String {
        let mut lines = vec![format!(
            "  BM25:     {}",
            if self.healthy { "healthy" } else { "degraded" }
        )];
        for index in &self.indexes {
            lines.push(format!("    {}: {}", index.name, index.state));
            if let Some(error) = &index.error {
                lines.push(format!("      Error: {error}"));
            }
        }
        if !self.healthy {
            lines.push(format!("    Repair: {}", self.repair_command));
        }
        lines.join("\n")
    }
}

fn verify_index(conn: &mut Client, name: &'static str) -> Bm25IndexHealth {
    let exists = conn
        .query_one("SELECT to_regclass($1::TEXT)::TEXT", &[&name])
        .and_then(|row| row.try_get::<_, Option<String>>(0));
    match exists {
        Ok(None) => {
            return index_health(
                name,
                "missing",
                Some("required BM25 index is missing; run PostgreSQL setup/migrations".into()),
            );
        }
        Err(error) => return verification_error(name, &error),
        Ok(Some(_)) => {}
    }

    match conn.query(
        "SELECT check_name, passed, details
         FROM pdb.verify_index($1::TEXT::regclass, on_error_stop => true)",
        &[&name],
    ) {
        Ok(rows) => {
            let checks = rows
                .into_iter()
                .map(|row| Bm25IndexCheck {
                    name: row.get("check_name"),
                    passed: row.get("passed"),
                    details: row.get("details"),
                })
                .collect::<Vec<_>>();
            if checks.is_empty() {
                return index_health(
                    name,
                    "error",
                    Some("pdb.verify_index returned no verification checks".into()),
                );
            }
            let healthy = checks.iter().all(|check| check.passed);
            let error = (!healthy).then(|| {
                checks
                    .iter()
                    .filter(|check| !check.passed)
                    .map(|check| check.details.as_deref().unwrap_or(&check.name))
                    .collect::<Vec<_>>()
                    .join("; ")
            });
            Bm25IndexHealth {
                name,
                state: if healthy { "healthy" } else { "damaged" },
                repaired: false,
                checks,
                error,
            }
        }
        Err(error) => verification_error(name, &error),
    }
}

fn verification_error(name: &'static str, error: &postgres::Error) -> Bm25IndexHealth {
    let state = if error
        .code()
        .is_some_and(|code| CORRUPTION_SQLSTATES.contains(&code.code()))
    {
        "damaged"
    } else {
        "error"
    };
    index_health(name, state, Some(postgres_errors::message(error)))
}

fn index_health(name: &'static str, state: &'static str, error: Option<String>) -> Bm25IndexHealth {
    Bm25IndexHealth {
        name,
        state,
        repaired: false,
        checks: Vec::new(),
        error,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn degraded_text_includes_repair_command_and_error() {
        let health = Bm25Health {
            healthy: false,
            repair_command: REPAIR_COMMAND,
            indexes: vec![index_health(
                REQUIRED_INDEXES[0],
                "damaged",
                Some("invalid chunk style tag: 254".into()),
            )],
        };

        let text = health.render_text();
        assert!(text.contains("invalid chunk style tag: 254"));
        assert!(text.contains(REPAIR_COMMAND));
    }
}

File: src/gobby/code_index/bm25_health.py
"""Health checks and targeted recovery for pg_search BM25 indexes."""

from __future__ import annotations

import time
from contextlib import nullcontext
from typing import Any

import psycopg
from psycopg import sql

BM25_INDEXES = (
    "public.code_symbols_search_bm25",
    "public.code_content_search_bm25",
)
BM25_REPAIR_COMMAND = "gobby postgres repair-code-index"

_REPAIR_LOCK_NAME = "gobby:code-index-bm25-repair"
_CORRUPTION_SQLSTATES = {"XX000", "XX001", "XX002"}


def unavailable_bm25_status(error: str) -> dict[str, Any]:
    """Return the stable degraded payload when PostgreSQL cannot be inspected."""
    return _status_payload(
        [_index_payload(name, state="error", error=error) for name in BM25_INDEXES]
    )


def verify_bm25_indexes(conn: Any) -> dict[str, Any]:
    """Verify every required BM25 index without mutating it."""
    return _status_payload([_verify_index(conn, name) for name in _required_index_names(conn)])


def repair_bm25_indexes(
    dsn: str,
    *,
    timeout_seconds: float = 900,
    connect_timeout: int = 5,
    lock_poll_seconds: float = 0.25,
) -> dict[str, Any]:
    """Verify and selectively rebuild damaged BM25 indexes."""
    try:
        with psycopg.connect(
            dsn,
            connect_timeout=connect_timeout,
            autocommit=True,
        ) as conn:
            conn.execute(
                "SELECT set_config('statement_timeout', %s, false)",
                (f"{max(1, int(timeout_seconds * 1000))}ms",),
            )
            if not _acquire_repair_lock(
                conn,
                timeout_seconds=timeout_seconds,
                poll_seconds=lock_poll_seconds,
            ):
                status = verify_bm25_indexes(conn)
                if status["healthy"]:
                    return status
                return _with_status_error(
                    status,
                    f"timed out waiting for BM25 repair lock after {timeout_seconds:g}s",
                )

            try:
                before = verify_bm25_indexes(conn)
                damaged = {item["name"] for item in before["indexes"] if item["state"] == "damaged"}
                rebuild_errors: dict[str, str] = {}
                for name in damaged:
                    try:
                        conn.execute(
                            sql.SQL("REINDEX INDEX {}").format(_qualified_identifier(name))
                        )
                    except psycopg.Error as exc:
                        rebuild_errors[name] = _postgres_error(exc)

                after = verify_bm25_indexes(conn)
                for item in after["indexes"]:
                    name = item["name"]
                    item["repaired"] = name in damaged and item["state"] == "healthy"
                    if name in rebuild_errors:
                        item["state"] = "error"
                        item["error"] = f"REINDEX failed: {rebuild_errors[name]}"
                after["healthy"] = all(item["state"] == "healthy" for item in after["indexes"])
                return after
            finally:
                conn.execute(
                    "SELECT pg_advisory_unlock(hashtextextended(%s, 0))",
                    (_REPAIR_LOCK_NAME,),
                )
    except psycopg.Error as exc:
        return unavailable_bm25_status(_postgres_error(exc))


def render_bm25_status(status: dict[str, Any]) -> list[str]:
    """Render the code-index portion of PostgreSQL status output."""
    lines = [f"Code index:  {'healthy' if status.get('healthy') else 'degraded'}"]
    for item in status.get("indexes", []):
        suffix = " (repaired)" if item.get("repaired") else ""
        lines.append(f"  {item['name']}: {item['state']}{suffix}")
        if item.get("error"):
            lines.append(f"    Error: {item['error']}")
    if not status.get("healthy"):
        lines.append(f"  Repair: {status.get('repair_command', BM25_REPAIR_COMMAND)}")
    return lines


def _verify_index(conn: Any, name: str) -> dict[str, Any]:
    try:
        transaction = getattr(conn, "transaction", None)
        with transaction() if callable(transaction) else nullcontext():
            row = conn.execute("SELECT to_regclass(%s)::text", (name,)).fetchone()
            if row is None or row[0] is None:
                return _index_payload(
                    name,
                    state="missing",
                    error="required BM25 index is missing; run PostgreSQL setup/migrations",
                )
            rows = conn.execute(
                """
                SELECT check_name, passed, details
                FROM pdb.verify_index(%s::regclass, on_error_stop => true)
                """,
                (name,),
            ).fetchall()
    except psycopg.Error as exc:
        state = "damaged" if exc.sqlstate in _CORRUPTION_SQLSTATES else "error"
        return _index_payload(name, state=state, error=_postgres_error(exc))

    checks = [
        {
            "name": str(check_name),
            "passed": bool(passed),
            "details": None if details is None else str(details),
        }
        for check_name, passed, details in rows
    ]
    if not checks:
        return _index_payload(
            name,
            state="error",
            error="pdb.verify_index returned no verification checks",
        )
    state = "healthy" if all(check["passed"] for check in checks) else "damaged"
    error = None
    if state == "damaged":
        error = "; ".join(
            str(check["details"] or check["name"]) for check in checks if not check["passed"]
        )
    return _index_payload(name, state=state, checks=checks, error=error)


def _acquire_repair_lock(
    conn: Any,
    *,
    timeout_seconds: float,
    poll_seconds: float,
) -> bool:
    deadline = time.monotonic() + max(0, timeout_seconds)
    while True:
        row = conn.execute(
            "SELECT pg_try_advisory_lock(hashtextextended(%s, 0))",
            (_REPAIR_LOCK_NAME,),
        ).fetchone()
        if row and bool(row[0]):
            return True
        if time.monotonic() >= deadline:
            return False
        time.sleep(min(poll_seconds, max(0, deadline - time.monotonic())))


def _qualified_identifier(name: str) -> sql.Identifier:
    schema, index = name.split(".", 1)
    return sql.Identifier(schema, index)


def _required_index_names(conn: Any) -> tuple[str, ...]:
    """Qualify required indexes with the connection's active schema."""
    row = conn.execute("SELECT current_schema()").fetchone()
    schema = str(row[0]) if row and row[0] else "public"
    return tuple(f"{schema}.{name.rsplit('.', 1)[1]}" for name in BM25_INDEXES)


def _index_payload(
    name: str,
    *,
    state: str,
    checks: list[dict[str, Any]] | None = None,
    error: str | None = None,
) -> dict[str, Any]:
    return {
        "name": name,
        "state": state,
        "repaired": False,
        "checks": checks or [],
        "error": error,
    }


def _status_payload(indexes: list[dict[str, Any]]) -> dict[str, Any]:
    return {
        "healthy": all(item["state"] == "healthy" for item in indexes),
        "repair_command": BM25_REPAIR_COMMAND,
        "indexes": indexes,
    }


def _with_status_error(status: dict[str, Any], error: str) -> dict[str, Any]:
    for item in status["indexes"]:
        if item["state"] != "healthy":
            item["error"] = f"{item['error']}; {error}" if item["error"] else error
    return status


def _postgres_error(exc: psycopg.Error) -> str:
    parts: list[str] = []
    diag = getattr(exc, "diag", None)
    for value in (
        getattr(diag, "message_primary", None),
        getattr(diag, "message_detail", None),
        getattr(diag, "message_hint", None),
    ):
        if value and value not in parts:
            parts.append(str(value))
    fallback = str(exc).strip()
    if fallback and fallback not in parts:
        parts.append(fallback)
    return ": ".join(parts) or type(exc).__name__
